### PR TITLE
[clang][analyzer] Support `ownership_{returns,takes}` attributes

### LINKED ISSUE #76861: [Clang SA]: custom malloc/free checker
Hi,

today unix.malloc can be customized with attribute with
__attribute((ownership_returns(malloc, size_pos))
__attribute((ownership_takes(malloc, free_ptr_pos)))

But in some system there can be several allocator. I could be usefull to check that malloc_1 are free with free_1, malloc_2, free with free_2.

Also how do you model custom realloc ?

### PR DESCRIPTION
**Pull request by pskrgag**
Add support for checking mismatched ownership_returns/ownership_takes attributes.

Closes #76861

### CODE REVIEW

**Comment by pskrgag** [comment]
Hello @steakhal,

could you, please, take a look? I just found linked issue quite interesting and decided to give it a try. I am new to llvm world, so I would really appreciate your review

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:119]: I'd suggest marking this `explicit`.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:131]: I don't see why a `return std::tie(_kind, _customName) == std::tie(Other._kind, Other._customName)` wouldn't work.
I find that more conventional to the one you proposed.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:136]: Can we drop this?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:138]: Could you define this for the class type instead?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:145]: I think it should be fine to have these fields as public members, given the complexity of this class. This is basically a std::pair. Feel free to use `struct` instead of `class`.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:229]: This Profile wouldn't take the custom allocation function name into account.
I'd suggest letting the Family define its own Profile method and just pass the ID to it.
That is the canonical way of doing this.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1709]: You could do this in the ctor of `AllocationFamily`. That would ensure that it's used like that consistently across all sites.

I've seen a variable name `Family` for representing the same value before. Have you considered using the same variable name to boost consistency and readability across the patch?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1924]: You could switch this into an early-return to reduce indentation for the rest of the code.
Same applies to the `if (const CallExpr *CE = dyn_cast<CallExpr>(E))` a few lines above.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1936]: You use single quoted char and double quoted char. I just highlight this inconsistency.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:2006]: I'd suggest swapping this `llvm_unreachable` to an `assert(false && "not a deallocation expression"); return;` as this code is probably not in the hot path, thus the use of `llvm_unreachable` is unjustified.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:2032]: Same goes here, use a regular `assert` and `return`.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:3627]: This isn't part of your current patch, but I'd suggest turning this into an assert and return too.

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (steakhal) [inline clang/test/Analysis/MismatchedDeallocator-checker-test.mm:22]: I'm surprised to see that a function can have two different `ownership_takes` attributes.
Are you sure that this is the case? If so, why is that?

Have you considered adding some documentation to the attributes page, and to the Malloc checker to highlight this attribute?

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (steakhal) [inline clang/test/Analysis/MismatchedDeallocator-checker-test.mm:87]: It's interesting to see that in a single message we have `my_malloc1()` unquoted; `malloc1` quoted; `my_free()` unquoted; finally `malloc` quoted.
Do you think it's worth it to harmonize these?
Also notice that some of these has parens in the end, some don't.

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (steakhal) [inline clang/test/Analysis/MismatchedDeallocator-checker-test.mm:17]: What happens if you have a forward declaration to a malloc function without the attribute, then later have another redeclaration with the `ownership_returns` attribute?
A similar case to test would be the opposite: 1st declaration has the attribute, the second redeclaration would not have the attribute.
I suspect that the FunctionDecl of the callees will refer to the last seen declaration of the function, thus, will not recognize the attribute for the second case I mentioned here.
Could you test these cases?

**Review by steakhal — CHANGES_REQUESTED** [review]
Looks pretty good. Thanks for working on the issue.
I didn't see any major blocking issue with the PR, so good job on that!

Overall, be sure to follow the llvm [coding style](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly), including using upper case variable names.

Thanks again for working on this :)

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (pskrgag) [reply clang/test/Analysis/MismatchedDeallocator-checker-test.mm:22]: I also find this quite strange, but can't say if there are any real-world examples of function that takes ownership of 2 different classes.

Maybe it would be better to just disallow 2 different classes?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (pskrgag) [reply clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:119]: Marking with `explicit` would cause compile errors for code like `family == AF_Smth`. We could wrap all `AF_*` with direct constructor call, but, I think, it would be less clean. What do you think?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (pskrgag) [reply clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:2006]: I think, all `llvm_unreachable` s  in this file could be turned in `assert` , since they just check sanity of calling contract. Will clean up all of them in separate patch, thanks!

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (steakhal) [reply clang/test/Analysis/MismatchedDeallocator-checker-test.mm:22]: To me, it would make more sense to allow at most one of this attribute to be present.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [reply clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:119]: I'd prefer having this explicit.

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (steakhal) [reply clang/test/Analysis/MismatchedDeallocator-checker-test.mm:87]: Ah, I see. The phrasing wasn't clear to me. Maybe because we don't have a clear spec for this attribute, thus it's difficult to make out things that have special meaning.
I think we should quote function names, yes.

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (pskrgag) [reply clang/test/Analysis/MismatchedDeallocator-checker-test.mm:22]: Disallowed 2 ownership_takes with different classes

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (pskrgag) [reply clang/test/Analysis/MismatchedDeallocator-checker-test.mm:87]: So I added quotes around all functions and operators. Output looks a way more consistent. Thanks!

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (pskrgag) [reply clang/test/Analysis/MismatchedDeallocator-checker-test.mm:17]: Seems like it works fine. At least, new test case passes. I guess, clang picks the most specific declaration

File: clang/test/Sema/attr-ownership.c
Comment (steakhal) [inline clang/test/Sema/attr-ownership.c:30]: What happens if I want to redeclare the function with a different "ownership_takes" attribute in a subsequent line? Will it detect the conflict?
```
int f20(void *) __attribute__((ownership_takes(foo, 1)));
int f20(void *) __attribute__((ownership_takes(foo1, 1))); // redecl with different attr
```

File: clang/test/Analysis/MismatchedDeallocator-checker-test.mm
Comment (steakhal) [reply clang/test/Analysis/MismatchedDeallocator-checker-test.mm:17]: Interesting.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:148]: I think you could also do this unconditionally.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:124]: I'd prefer to have a message here to be able to immediately tell what this assert is about when fires.
```suggestion
    assert((kind != AF_Custom || name.has_value()) && "Custom family must specify also the name");
```

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:122]: Please use upper camel case for variable names. Shadowing is not a problem for constructor parameters and init-lists, and within the ctor body the semantics are anyways well-defined.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:132]: Probably I don't understand the semantics of this one.
I was expecting that the `CustomName` field is only used when the `Kind` is `AF_Custom`. Yet, here it appears to be used even for `Kind == AF_Malloc`.

What I originally thought that when `AF_Custom` is specified with `"malloc"`, then it should consider the case as `AF_Malloc` instead.
Could you elaborate here?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1160]: Now I get what you meant when pushed back on adding `explicit`. I didn't expect this to cause such a widespread effect. Sorry. Now that you made all the necessary changes, we can keep it as-is. You decide if you prefer reverting these back. I'm good either way.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1535]:
```suggestion
  } else {
    assert(false && "not a CallExpr or CXXNewExpr");
    return nullptr;
  }
```
Just to be on the safe side and we have a NDEBUG build (with assertions disabled), let's avoid undefined behavior by safely backing off.

The same principle should apply to all the other cases when previously we had an `llvm_unreachable`.
This is what I wanted to imply by `assert and return`.

**Review by steakhal — COMMENTED** [review]
It looks great! I only had one more thing to test.
Besides that it would be nice to advertise this in the `clang/docs/ReleaseNotes.rst`.
Usually, from that release notes we also link to the checker or the relevant doc pages - in this case it would be nice to link to the docs of this attribute - which does not really elaborate on the attribute and the semantics.

Do you want to extend the scope of this PR to add some minimal docs to the attribute?
If not, that's also fine, we will create a separate ticket for adding them later.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (pskrgag) [reply clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1159]: I like current variant more. Just because it makes code more straightforward, I believe

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (pskrgag) [reply clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp]: Sorry, I over-engineered things. 

I redo a bit. Now callers pass `AF_Custom` with `customName` set, instead of `AF_Malloc`

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (pskrgag) [reply clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:147]: Why? If it's not custom class, then unwrapping `CustomName` would cause an exception, isn't it?

File: clang/test/Sema/attr-ownership.c
Comment (pskrgag) [reply clang/test/Sema/attr-ownership.c:30]: Hmmmmm.... It does not detect. As well as for other attributes. 

I think, we should fix it, but maybe all of them in scope of separate PR?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [reply clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:147]: Ah indeed. nvm.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1928]: When I advocated for early-returns, I wanted something like this:
```c++
const auto *CE = dyn_cast<CallExpr>(E);
if (!CE)
  return;
const FunctionDecl *FD = CE->getDirectCallee();
if (!FD)
  return;
  ...
```
On the same token, the `FD->hasAttrs()` appears to be not strictly necessary as in that case the loop simply wouldn't take any iterations at all, right?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1940]:
```suggestion
      if (I->getOwnKind() != OwnershipAttr::Takes)
```
It seems like inlining this variable would make it more readable actually.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1944]:
```suggestion
      os << ", which takes ownership of '" << I->getModule()->getName() << '\'';
```
This way it would even fit into a single line.

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:1936]:
```suggestion
    // Only one ownership_takes attribute is allowed.
```

**Review by steakhal — APPROVED** [review]
I think it looks great.
Only a few more stylistic remarks are left, but it should be already in a good shape functional-wise. I'll let you decide if you wanna apply my suggestions or not.

One more remark - this time w.r.t. the workflow. Usually, GitHub PRs prefer "merges" over "force-pushes". Whenever you force-push, all the inline remarks could get lost, as it fails to track the lines where the comments were made. Thus, when working with PRs, one usually just "merges" the tracked branch, and just builds more and more commits on top of your PR branch. Once the review is done, one is free to destroy the history and squash the commits, or do some reordering - because there are no more pending comments or remarks left.

**Comment by pskrgag** [comment]
Thank you for review!

I've applied your suggestions and fixed test failures. Messed up a bit with shadowing in constructor. 

> Usually, GitHub PRs prefer "merges" over "force-pushes". Whenever you force-push, all the inline remarks could get lost, as it fails to track the lines where the comments were made

Oh, ok. I've never really worked with gh workflow, so didn't know that. Thank you for guidance!

**Comment by steakhal** [comment]
FYI This PR is still not mentioned in the `clang/docs/ReleaseNotes.rst`.
We should have two entries there:
 1) In [Attribute Changes in Clang](https://github.com/llvm/llvm-project/blob/main/clang/docs/ReleaseNotes.rst#id34), mentioning that calls can have at most one `ownership_return` attribute.
 2) In [Static Analyzer > New features](https://github.com/llvm/llvm-project/blob/main/clang/docs/ReleaseNotes.rst#id47), that the MallocChecker now checks this attribute, and you can read the docs for the checker, and the attribute at the links.

**Comment by pskrgag** [comment]
> and you can read the docs for the checker, and the attribute at the links.

I can't find any docs for this attribute. As I mentioned, I will fill new issues to fix couple of frontend issues and after that we can write down correct semantics of these attrs.

Do you mean that link https://clang.llvm.org/docs/AttributeReference.html#ownership-holds-ownership-returns-ownership-takes?

**Comment by steakhal** [comment]
> > and you can read the docs for the checker, and the attribute at the links.
> 
> I can't find any docs for this attribute. As I mentioned, I will fill new issues to fix couple of frontend issues and after that we can write down correct semantics of these attrs.

Thats fine.

> Do you mean that link https://clang.llvm.org/docs/AttributeReference.html#ownership-holds-ownership-returns-ownership-takes?

Yes. Its generated from a tablegen file. Thats why you probably didnt find it.
Look for files with the "td" extension, and  "attr" in name. May have a different casing though. Watch out.

**Comment by pskrgag** [comment]
Rebased on top of d89f3e8df3160b3afc07bc742c81aa4738ea9646

**Review by steakhal — APPROVED** [review]

**Comment by steakhal** [comment]
Should I merge this PR, or you want to wait some?

**Comment by pskrgag** [comment]
I think, it's good time to merge, since v19 release branch was created.

If you want to wait for others to review, it's also fine.

**Pull request merged**
